### PR TITLE
Add support for copy files phase

### PR DESCRIPTION
**Pull request by hebertialmeida**
Resolves https://github.com/tuist/tuist/issues/2075

### Short description 📝

> This adds the ability to copy files as a copy files phase.

### Solution 📦

> Copy the files and create references similiar to resources and add a `PBXCopyFilesBuildPhase`. One issue is that the packages are not "recognized" on Tuist, so I had to add some validations to check the UTI of a file in order to add the custom packages as whole file instead of its internal file list.

### Implementation 👩‍💻👨‍💻

> I've reused the implementation of resources for some steps or I've tried to follow the same pattern. One thing I considered is to add a `static func` for the `CopyFilesAction` instead of using `init`, just like `TargetActions`.

Sample usage:
```swift
Target(
    name: "App",
    platform: .macOS,
    product: .app,
    bundleId: "io.tuist.App",
    copyFiles: [
        .sharedSupport(
            name: "Copy Templates",
            subpath: "Templates",
            files: ["/SharedSupport/Templates/**"]
        ),
        .resources(
            name: "Copy Fonts",
            subpath: "Fonts",
            files: ["/Fonts/**"]
        )
    ]
)
```

- [x] Validate api design
- [x] Add tests
- [x] Add fixture
- [x] Update documentation
- [x] Update CHANGELOG

### CODE REVIEW

File: Sources/ProjectDescription/CopyFilesAction.swift
Comment (fortmarek) [inline Sources/ProjectDescription/CopyFilesAction.swift:10]: where are these numbers coming from? Are we sure that this list covers all use-cases?

File: Sources/ProjectDescription/CopyFilesAction.swift
Comment (fortmarek) [inline Sources/ProjectDescription/CopyFilesAction.swift:3]: I am not sure I would call it `CopyFilesAction` -> I'd still lean to `BundledFiles` as that is more descriptive of what the user wants to achieve.

File: Sources/ProjectDescription/CopyFilesAction.swift
Comment (fortmarek) [inline Sources/ProjectDescription/CopyFilesAction.swift:4]: I don't think we want to let the users name the action, let's do that automatically

File: Sources/ProjectDescription/Target.swift
Comment (fortmarek) [inline Sources/ProjectDescription/Target.swift:44]: I would not call it a phase, something like `Files to be bundled with application` would be better, I think

File: Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift
Comment (fortmarek) [inline Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift:267]: It should probably be `BuildPhases` as this can produce multiple of them

File: Sources/ProjectDescription/CopyFilesAction.swift
Comment (fortmarek) [reply Sources/ProjectDescription/CopyFilesAction.swift:10]: Also - do we need these raw values in `ProjectDescription`?

File: Sources/TuistSupport/Extensions/AbsolutePath+Extras.swift
Comment (fortmarek) [inline Sources/TuistSupport/Extensions/AbsolutePath+Extras.swift:26]: could you give me a little more context of why this is needed? What's the user input (`files` in `CopyFilesAction`) that makes this necessary?

**Review by fortmarek — COMMENTED** [review]
Thanks for doing this! I have some initial remarks, maybe adding a [fixture](https://tuist.io/docs/contribution/testing-strategy/#acceptance-tests) for the new API would help me understand some steps you have taken (especially the UTI handling)

File: Sources/ProjectDescription/CopyFilesAction.swift
Comment (hebertialmeida) [reply Sources/ProjectDescription/CopyFilesAction.swift:10]: These values map directly to [PBXCopyFilesBuildPhase.swift#L5](https://github.com/tuist/XcodeProj/blob/main/Sources/XcodeProj/Objects/BuildPhase/PBXCopyFilesBuildPhase.swift#L5) on `func generateCopyFilesBuildPhase(path:` and it covers all the use-cases. You are right we don't need the raw values in `ProjectDescription`. I'm going to update that.

**Comment by hebertialmeida** [comment]
Adding a little bit more of context to this. 

On iOS apps everything is bundled under `Copy Bundle Resources` and the files you add for the app will be inside the app bundle. On macOS apps things are a little bit different, for example custom fonts have to be added to a custom folder via copy files phase and referred on .plist file using the key `ATSApplicationFontsPath`, not as simple as on iOS. Adding support or that will enable these apps to copy files/executables to the proper paths, things like themes, templates, fonts...

Many times these files are packages, basically a folder with a custom extension like `.pages`, `.sketch` but they are recognized by the OS as a single file because these apps register a UTI for that extension.

That's why I believe having the `name` property is important so you see a can organized list and make sense of what is each phase. In addition to that I think naming it `CopyFilesPhase` makes more sense than `BundledFiles`. In some places I've used the word `Action` but that's I'm not sure about that...

This is an example of my app:
image

Apple Pages also do the same thing:
image

https://help.apple.com/xcode/mac/10.2/#/dev50bab713d

File: Sources/ProjectDescription/Target.swift
Comment (hebertialmeida) [reply Sources/ProjectDescription/Target.swift:44]: I've added a comment with more context, see if that makes sense for you...

File: Sources/ProjectDescription/CopyFilesAction.swift
Comment (hebertialmeida) [reply Sources/ProjectDescription/CopyFilesAction.swift:4]: I've added a comment with more context, but basically this is important so we can organize the phases and understand what is going on, this name is visible.

File: Sources/ProjectDescription/CopyFilesAction.swift
Comment (hebertialmeida) [reply Sources/ProjectDescription/CopyFilesAction.swift:3]: I've added a comment with more context, see if that makes sense for you... I'm leaning towards `CopyFilesPhase` or `CopyFilesAction` but I think it's important to mention Copy Files, as these are not really Bundled files...

**Comment by hebertialmeida** [comment]
> Thanks for doing this! I have some initial remarks, maybe adding a [fixture](https://tuist.io/docs/contribution/testing-strategy/#acceptance-tests) for the new API would help me understand some steps you have taken (especially the UTI handling)

Will do that!

File: Sources/ProjectDescription/Target.swift
Comment (hebertialmeida) [reply Sources/ProjectDescription/Target.swift:44]: But agreed that this need better clarification for documentation purpose.

File: Sources/TuistSupport/Extensions/AbsolutePath+Extras.swift
Comment (hebertialmeida) [reply Sources/TuistSupport/Extensions/AbsolutePath+Extras.swift:26]: This can have a better name, but not sure what to name it...

Given the input input:
```
/User/name/project/Templates/yellow.template/meta.json
/User/name/project/Templates/yellow.template/image1.png
/User/name/project/Templates/yellow.template/image2.png
/User/name/project/Templates/blue.template/meta.json
/User/name/project/Templates/blue.template/image1.png
/User/name/project/Templates/blue.template/image2.png
/User/name/project/Fonts/somefont.ttf
```

This is the output:
```
/User/name/project/Templates/yellow.template
/User/name/project/Templates/blue.template
/User/name/project/Fonts/somefont.ttf
```

Comment (hebertialmeida) [reply Sources/TuistSupport/Extensions/AbsolutePath+Extras.swift:26]: It will basically clean the list for packages.

File: Sources/ProjectDescription/CopyFilesAction.swift
Comment (pepibumur) [reply Sources/ProjectDescription/CopyFilesAction.swift:3]: Yeah, bundles are used for something else in Xcode so I'd rather not use it. I like `CopyFilesAction`, we also use actions to refer to script build phases so I think it's consistent with what we've done previously.

File: Sources/ProjectDescription/CopyFilesAction.swift
Comment (pepibumur) [inline Sources/ProjectDescription/CopyFilesAction.swift:7]: I'd document every public element in this file. That way, developers can get inline documentation when editing the manifest with `tuist edit`.

File: Sources/ProjectDescription/CopyFilesAction.swift
Comment (pepibumur) [inline Sources/ProjectDescription/CopyFilesAction.swift:23]: I'd keep this method private, and add a static initializer:

```swift
let target = Target(copyFiles: .plugins(name: "Copy X", files: [])
let target = Target(copyFiles: .frameworks(name: "Copy X", files: [])

// vs
let target = Target(copyFiles: .init(name: "Copy X", destination: .plugins, subpath: ""))
let target = Target(copyFiles: CopyFilesAction(name: "Copy X", destination: .plugins, subpath: ""))
```
Notice how the first one removes the redundancy that `copyFiles: CopyFilesActions` would introduce and I think it makes the API more expressive.

File: Sources/ProjectDescription/CopyFilesAction.swift
Comment (pepibumur) [inline Sources/ProjectDescription/CopyFilesAction.swift:25]: Is `subpath` required always?

File: Sources/TuistCache/ContentHashing/CopyFilesContentHasher.swift
Comment (pepibumur) [inline Sources/TuistCache/ContentHashing/CopyFilesContentHasher.swift:24]: Not sure if it's possible in the case of copy files, but with scripts, we couldn't hash the files because paths can contain variables, e.g. ${SRCROOT}, that are resolved at build time.

File: Sources/TuistCache/ContentHashing/CopyFilesContentHasher.swift
Comment (pepibumur) [inline Sources/TuistCache/ContentHashing/CopyFilesContentHasher.swift:21]: I'd add a unit test for this one.

File: Sources/TuistCore/Models/CopyFilesAction.swift
Comment (pepibumur) [inline Sources/TuistCore/Models/CopyFilesAction.swift:11]: The fact that those values are the ones used in Xcode projects is not obvious. I'd rather have something like:

```swift
public enum Destination: Equatable {
  var xcodeprojValue: UInt {
    // switch 
  }
}
```

File: Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift
Comment (pepibumur) [inline Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift:267]: I'd add a unit test that ensures that the project contains the right copy files build phases.

File: Sources/TuistGenerator/Generator/ProjectFileElements.swift
Comment (pepibumur) [inline Sources/TuistGenerator/Generator/ProjectFileElements.swift:146]: Same here, I'd add a test to ensure that we generate file elements for the files referenced by the copy files.

File: Sources/TuistSupport/Extensions/AbsolutePath+Extras.swift
Comment (pepibumur) [reply Sources/TuistSupport/Extensions/AbsolutePath+Extras.swift:26]: Couldn't we pass the second list instead when declaring the list of files? Or is it because you want to use glob wildcards?

**Review by pepibumur — COMMENTED** [review]
It's looking great @hebertialmeida. Very solid work and consistent with the rest of the codebase 👏. I'm looking forward to seeing this feature on Tuist.
One more suggestion that I'd make is to make sure to update the CHANGELOG and the documentation website.

File: Sources/ProjectDescription/CopyFilesAction.swift
Comment (hebertialmeida) [reply Sources/ProjectDescription/CopyFilesAction.swift:25]: It's not on Xcode, making it optional

**Review by kwridan — COMMENTED** [review]
Nice work on this @hebertialmeida - thanks for adding this!

@fortmarek and @pepibumur have covered most points, one addition to help us test this locally and show case this feature is to introduce a new fixture in `fixtures/` that exercises the new additions.

File: Sources/TuistSupport/Extensions/AbsolutePath+Extras.swift
Comment (hebertialmeida) [reply Sources/TuistSupport/Extensions/AbsolutePath+Extras.swift:26]: Yeah, I thought about using glob wildcards just like for resources

File: Sources/TuistCache/ContentHashing/CopyFilesContentHasher.swift
Comment (hebertialmeida) [reply Sources/TuistCache/ContentHashing/CopyFilesContentHasher.swift:24]: I think for files this should be ok, like the list of resources.

**Comment by hebertialmeida** [comment]
I'm seeing the linter fail in lines I haven't touched, is that normal? Should I just fix it?

File: Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift
Comment (hebertialmeida) [reply Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift:283]: I've changed the `isPackage` check so it does not perform any disk check, instead it will check the UTI based of the file extension, that's reliable for testing. That is what `cleanPackages` uses to check and also that's only being executed once, just like `generateResourcesBuildFile` for removing files like .lproj and .xcassets.

Comment (kwridan) [reply Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift:283]: Awesome, thanks for the update!

The concern I had was if in inconsistency occurs between `ProjectFileElements` and `BuildPhaseGenerator`. `ProjectFileElements` currently constructs groups and elements based on the path. 

For example if it encounters the path:
`/Project/Templates/MyTemplate.template/manifest.json`

Given the project is in `/Project` the following will be created:

- Group `Templates`
- Group `Templates/MyTemplate.template`
- File Element `Templates/MyTemplate.template/manifest.json`

As such when `BuildPhaseGenerator` asks for a file element for `/Project/Templates/MyTemplate.template` it may fail seeing `/Project/Templates/MyTemplate.template` is a group. 

Interestingly though from testing this PR it doesn't seem we hit this issue - looks like the internal sorting of paths results in the package path `/Project/Templates/MyTemplate.template` being encountered before any of its nested files resulting in the file element for it getting created and subsequently paths for the nested files get dropped.

I guess it may become an issue later on if that behaviour changes - but at least now there is a fixture with acceptance tests that can help us catch regressions if anything changes 👍

**Review by kwridan — COMMENTED** [review]
> I'm seeing the linter fail in lines I haven't touched, is that normal? Should I just fix it?

I think the linter flags any issues on the file level rather than the diff - should be ok to fix it. There are a few rake tasks to run the same checks as CI

```sh
bundle exec rake style_correct
```

File: Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift
Comment (hebertialmeida) [reply Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift]: I got that! That makes sense, I've made the changes and now `cleanPackages` happens inside `CopyFilesAction+ManifestMapper` right after the globs are resolved and turned into paths.

**Review by kwridan — APPROVED** [review]
Thanks for all the updates @hebertialmeida - the PR looking great!

**Review by pepibumur — APPROVED** [review]
Thank you very much @hebertialmeida for addressing the comments. There's only one more thing and we can merge the PR afterward, adding documentation. Since we are changing the public API that users have access to, I'd recommend updating [this page](https://github.com/tuist/tuist/blob/main/website/markdown/docs/usage/project-description.mdx) to indicate that targets have now an attribute, and what the attribute is for.

**Comment by pepibumur** [comment]
@all-contributors add @hebertialmeida

**Comment by pepibumur** [comment]
@all-contributors add @hebertialmeida for code

**Comment by hebertialmeida** [comment]
@pepibumur @fortmarek @kwridan thanks for the guidance and @natanrolnik for helping me build the project! I have updated the documentation and will merge that after all checks pass.

**Comment by hebertialmeida** [comment]
I can't merge it, so please merge it when you can...

**Pull request merged**
